src/Npgsql.SourceGenerators/DatabaseInfoSourceGenerator.cs

```diff
@@ -0,0 +1,643 @@
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Scriban;
+using Scriban.Runtime;
+
+namespace Npgsql.SourceGenerators;
+
+[Generator]
+sealed class DatabaseInfoSourceGenerator : ISourceGenerator
+{
+    // Warning: Do not change to target typing
+    // See: https://github.com/dotnet/roslyn-analyzers/issues/5890#issuecomment-1046043775
+    // ReSharper disable once ArrangeObjectCreationWhenTypeEvident
+    static readonly DiagnosticDescriptor DatFileParserError = new DiagnosticDescriptor(
+        id: "NPGSQLDBINFOGEN001",
+        title: "Failed to parse .dat file",
+        messageFormat: "Syntax error in .dat file",
+        category: "DatabaseInfoSourceGenerator",
+        DiagnosticSeverity.Error,
+        isEnabledByDefault: true);
+
+    public void Initialize(GeneratorInitializationContext context) { }
+
+    public void Execute(GeneratorExecutionContext context)
+    {
+
+        List<DatFileObject>? pgType = null;
+        List<DatFileObject>? pgRange = null;
+        var noDatFile = true;
+        foreach (var file in context.AdditionalFiles)
+        {
+            if (Path.GetExtension(file.Path) != ".dat")
+                continue;
+
+            noDatFile = false;
+            var fileName = Path.GetFileName(file.Path);
+            try
+            {
+                switch (fileName)
+                {
+                case "pg_type.dat":
+                    pgType = ParseDatFile(file.GetText());
+                    break;
+                case "pg_range.dat":
+                    pgRange = ParseDatFile(file.GetText());
+                    break;
+                default:
+                    throw new InvalidOperationException($"This source generator is not designed to make use of the information in '{fileName}'. " +
+                                                        "You may want to open an issue at https://github.com/npgsql/npgsql/issues/new and describe " +
+                                                        "what you are trying to do.");
+                }
+            }
+            catch (DatFileFormatException e)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(DatFileParserError,
+                    Location.Create(file.Path, new(e.Position, 1),
+                        new(new(e.LineNumber - 1, e.LinePosition - 1), new(e.LineNumber - 1, e.LinePosition)))));
+
+                // We have detected and reported a syntax error in a .dat file so we just return
+                return;
+            }
+        }
+
+        // There is no .dat file in the additional files for this project.
+        // This probably means that the project just isn't interested in this source generator
+        if (noDatFile)
+            return;
+
+        if (pgType == null)
+            throw new ("This source generator needs at least a valid pg_type.dat file in AdditionalFiles to perform it's work");
```

Comment:
Is this a target-typed constructor? Which exception actually gets thrown, Exception?

No particular objection (since this is all internal stuff), but nit:

```suggestion
            throw new("This source generator needs at least a valid pg_type.dat file in AdditionalFiles to perform it's work");
```

Reply:
Yeah, that's just Exception.

Reply:
It's weird C# allows this!